pkg/node/node.go

```diff
@@ -179,13 +205,104 @@ func RefreshPeers(peers map[string]*Peer, logger log.Logger) (err error) {
 	wg.Wait()
 
 	// np contains all peers that had a reachable 26657. Simply add them into
-	// the peers map.
-	for _, p := range np.nodes {
-		peers[p.ID] = p
-	}
+	// the answer.
+	peersReachable = np.nodes
 	return
 }
 
+// getLatestBlockHeight retries GET /status until the node reports a correct latest block height
+func getLatestBlockHeight(peers map[string]*Peer, chainID string, logger log.Logger) (latestBlockHeight int64, err error) {
+	// Transform map into list of peers
+	vs := []*Peer{}
+	for _, v := range peers {
+		vs = append(vs, v)
+	}
+
+	for i := 0; i < len(vs); i++ {
+		latestBlockHeight, err = mustGetLatestBlockHeight(vs[i], chainID, logger)
+		if err == nil {
+			return latestBlockHeight, err
+		} else {
+			logger.Debug("Peer could not tell us the latest block height", "peer", vs[i].Address)
+		}
+	}
+	return 0, fmt.Errorf("update light roots: no peer could tell us the latest block height")
+}
+
+func mustGetLatestBlockHeight(peer *Peer, chainID string, logger log.Logger) (latestBlockHeight int64, err error) {
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
+	defer cancel()
+
+	client, e := Client(peer.Address)
+	if e != nil {
+		logger.Error("error creating tendermint client: %s", e)
+	}
+
+	for {
```

Comment:
nit: I'd just use a ticker here instead of manual `Sleep` calls 👍

Reply:
Interesting, didn't know about tickers before. How about now?